block/manager.go

```diff
@@ -642,65 +749,121 @@ func (m *Manager) BlockStoreRetrieveLoop(ctx context.Context) {
 				default:
 				}
 				// early validation to reject junk blocks
-				if !m.isUsingExpectedCentralizedSequencer(block) {
+				if !m.isUsingExpectedCentralizedSequencer(header) {
 					continue
 				}
-				m.logger.Debug("block retrieved from p2p block sync", "blockHeight", block.Height(), "daHeight", daHeight)
-				m.blockInCh <- NewBlockEvent{block, daHeight}
+				m.logger.Debug("header retrieved from p2p header sync", "headerHeight", header.Height(), "daHeight", daHeight)
+				m.headerInCh <- NewHeaderEvent{header, daHeight}
 			}
 		}
-		lastBlockStoreHeight = blockStoreHeight
+		lastHeaderStoreHeight = headerStoreHeight
 	}
 }
 
-func (m *Manager) getBlocksFromBlockStore(ctx context.Context, startHeight, endHeight uint64) ([]*types.Block, error) {
+// DataStoreRetrieveLoop is responsible for retrieving data from the Data Store.
+func (m *Manager) DataStoreRetrieveLoop(ctx context.Context) {
+	lastDataStoreHeight := uint64(0)
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-m.dataStoreCh:
+		}
+		dataStoreHeight := m.dataStore.Height()
+		if dataStoreHeight > lastDataStoreHeight {
+			data, err := m.getDataFromDataStore(ctx, lastDataStoreHeight+1, dataStoreHeight)
+			if err != nil {
+				m.logger.Error("failed to get data from Data Store", "lastDataStoreHeight", lastDataStoreHeight, "dataStoreHeight", dataStoreHeight, "errors", err.Error())
+				continue
+			}
+			daHeight := atomic.LoadUint64(&m.daHeight)
+			for _, d := range data {
+				// Check for shut down event prior to logging
+				// and sending header to dataInCh. The reason
+				// for checking for the shutdown event
+				// separately is due to the inconsistent nature
+				// of the select statement when multiple cases
+				// are satisfied.
+				select {
+				case <-ctx.Done():
+					return
+				default:
+				}
+				//TODO: remove junk if possible
+				m.logger.Debug("data retrieved from p2p data sync", "dataHeight", d.Metadata.Height, "daHeight", daHeight)
+				m.dataInCh <- NewDataEvent{d, daHeight}
+			}
+		}
+		lastDataStoreHeight = dataStoreHeight
+	}
+}
+
+func (m *Manager) getHeadersFromHeaderStore(ctx context.Context, startHeight, endHeight uint64) ([]*types.SignedHeader, error) {
+	if startHeight > endHeight {
+		return nil, fmt.Errorf("startHeight (%d) is greater than endHeight (%d)", startHeight, endHeight)
+	}
+	if startHeight == 0 {
+		startHeight++
+	}
+	headers := make([]*types.SignedHeader, endHeight-startHeight+1)
+	for i := startHeight; i <= endHeight; i++ {
+		header, err := m.headerStore.GetByHeight(ctx, i)
+		if err != nil {
+			return nil, err
+		}
+		headers[i-startHeight] = header
+	}
+	return headers, nil
+}
+
+func (m *Manager) getDataFromDataStore(ctx context.Context, startHeight, endHeight uint64) ([]*types.Data, error) {
 	if startHeight > endHeight {
 		return nil, fmt.Errorf("startHeight (%d) is greater than endHeight (%d)", startHeight, endHeight)
 	}
 	if startHeight == 0 {
 		startHeight++
 	}
```

Comment:
Similarly to `getHeaderFromDataStore` increment first, then check if `startHeight > endHeight`.